src/rgw/rgw_keystone_scope.h

```diff
@@ -0,0 +1,205 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
+// vim: ts=8 sw=2 smarttab ft=cpp
+
+#pragma once
+
+#include <optional>
+#include <string>
+#include <vector>
+
+#include "include/buffer.h"
+#include "include/encoding.h"
+
+class CephContext;
+
+namespace ceph {
+  class Formatter;
+}
+
+namespace rgw::keystone {
+
+// Forward declaration
+class TokenEnvelope;
+
+/**
+ * Keystone authentication scope information.
+ *
+ * This structure captures the OpenStack Keystone authentication context
+ * including project, user, and roles. It's used throughout the
+ * authentication and logging pipeline.
+ *
+ * The structure supports:
+ * - Binary encoding/decoding for RADOS backend (ops log storage)
+ * - JSON formatting for file/socket backend (ops log output)
+ * - Granular control via configuration flags (include_user, include_roles)
+ */
+struct ScopeInfo {
+  /**
+   * Keystone domain information.
+   * Domains provide namespace isolation in Keystone.
+   */
+  struct domain_t {
+    std::string id;
+    std::string name;
+
+    void encode(bufferlist &bl) const {
+      ENCODE_START(1, 1, bl);
+      encode(id, bl);
+      encode(name, bl);
+      ENCODE_FINISH(bl);
+    }
+    void decode(bufferlist::const_iterator &p) {
+      DECODE_START(1, p);
+      decode(id, p);
+      decode(name, p);
+      DECODE_FINISH(p);
+    }
+  };
+
+  /**
+   * Keystone project (tenant) information.
+   * Projects are the primary authorization scope in Keystone.
+   */
+  struct project_t {
+    std::string id;
+    std::string name;
+    domain_t domain;
+
+    void encode(bufferlist &bl) const {
+      ENCODE_START(1, 1, bl);
+      encode(id, bl);
+      encode(name, bl);
+      domain.encode(bl);
```

Comment:
please prefer the free functions `encode(domain, bl)` and `decode(domain, p)` for consistency with builtin/std types